src/workerd/api/sockets.c++

```diff
@@ -346,9 +352,28 @@ jsg::Ref<Socket> Socket::startTls(jsg::Lock& js, jsg::Optional<TlsOptions> tlsOp
     // that's the case.
     JSG_REQUIRE(
         *tlsStarter != kj::none, TypeError, "The request has finished before startTls completed.");
-    auto secureStream = KJ_ASSERT_NONNULL(*tlsStarter)(acceptedHostname)
+
+    auto& context = IoContext::current();
+    auto tlsStarterResult =
+        context.awaitIo(js, KJ_ASSERT_NONNULL(*tlsStarter)(acceptedHostname))
+            .then(js,
```

Comment:
`context.awaitIo(js, promise).then(func)` should almost always be changed to `context.awaitIo(js, promise, func)` instead, as this limits the lifetime of `func` to the I/O operation and allows it to capture I/O objects. If you use `.then()`, then the func's lifetime is governed by garbage collection, which is almost always worse.

That said, in this particular case, your captures look safe as you aren't capturing any I/O objects.

However, you _are_ capturing GC objects which could create cycles, and you are not using `JSG_VISITABLE_LAMBDA` to make the lambdas GC-visitable, so this could lead to memory leaks. Again, this is a non-issue if you pass the func directly to `awaitIo`.

Reply:
oh good catch! There's actually a `JSG_VISITABLE_LAMBDA necessary higher up in this function (in the `writable->flush(js).then(js, ...` call since that lambda ends up capturing the `jsg::Promise::Resolver` and the opened promise most likely captures a reference to itself in it's continuations.